### PR TITLE
net/metrics: initial commit

### PR DESCRIPTION
**Pull request by danrl**
Add metrics.lua, a Prometheus-compatible node exporter for OpenWrt/LEDE.

Signed-off-by: Dan Luedtke <mail@danrl.com>
 
Compile tested: LEDE latest, x86_64
Run tested: LEDE latest (and bi-daily snapshots from last month), x86_64

Description:

This is based on `metrics.lua` by @lyda 
(Source: https://gist.github.com/lyda/ba33d229a1c01fc5e445)

I only added the UCI interface to it to adjust listening port and restrict listening address.
Have been running this for a while now to produce graphics like this:

Asking @lyda for comments.

### CODE REVIEW

**Comment by lyda** [comment]
Oh nice, thanks!

**Comment by Ansuel** [comment]
What are we waiting for to add this to the tree?

**Comment by danrl** [comment]
I don't know :) We just need someone with write access to adopt this one.

**Comment by danrl** [comment]
Is there anything that blocks this and that I am overlooking right now?

File: net/metrics/Makefile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  TITLE:=Prometheus-compatible Metrics Exporter
+  PKGARCH:=all
+  MAINTAINER:=Dan Luedtke <mail@danrl.com>
+endef
+
+define Package/metrics/description
+  A Prometheus-compatible metrics exporter for monitoring and statistics.
+endef
+
+define Package/acme/conffiles
```
Comment (hnyman) [inline net/metrics/Makefile]: You should correct the package name here... this is not "acme"

Comment (danrl) [inline net/metrics/Makefile]: Done. Used ACME as template. Build still running. I'll let you know about the results.

**Comment by danrl** [comment]
Let me explain how node exporter for Prometheus work. I see now why there is confusion. The image I posted was from my Prometheus/Grafana dashboard. Prometheus is a time-series database that is popular for monitoring system metrics. People often run Grafana on top of it to get even fancier visualisations, such as the one you can see in the screenshot. Prometheus *pulls* the data from systems, like in the old days, when we used SNMP :) Prometheus pulls metrics via HTTP or HTTPS in a very simple format. There are other time-series database solutions that use *push* as paradigma, and there is generally a heated debate over which one is the better as soon as there is more than one expert in the room. I suggest a cold beer and a lot of patience, should you ever end up in such a situation.

So, what this little fellow (metrics.lua) does, is that it provides a webserver on `listen_port` (prometheus default is 9100) and on `listen_address`. For `listen_address` I chose ::1, as I refuse to write non-IPv6 capable software and defaults. I am sick of debugging legacy-first systems. I respect that others prefer what they already know and what they experienced to be reliable and stable. However, the script listens for HTTP GET requests for resource "/metrics" and serves system stats once the request was made. Then it waits for the next requests. It is common practice to pull metrics once per minute in many Prometheus setups. The data aggregated over the time series is then often used by Grafana to display almost real-time insight into the system's metrics and derived thereof: the system's health.


TL;DR: This package allows routers to be monitored by Prometheus. For security reasons it does only listen on localhost after installation. Admins have to configure `listen_address` to reflect their security policies.

Example: I have a setup where *metrics* listens on an address that is assigned to a WireGuard interface. Thus all metrics data is encrypted by the outer tunnel. All monitored systems have a WireGuard interface where the other end is on the Prometheus server which is hosted in a datacenter.

Thank you for reviewing the code. Please let me know if you need further information to proceed.

**Comment by danrl** [comment]
Build successful. Test successful.

Example:
````
root@lede-dev:~# nc ::1 9100
GET /metrics HTTP/1.0
HTTP/1.1 200 OK
Server: lua-metrics
Content-Type: text/plain; version=0.0.4

# TYPE node_boot_time gauge
node_boot_time 1.48854e+09
# TYPE node_context_switches counter
node_context_switches 17596
# TYPE node_cpu counter
node_cpu{cpu="cpu0",mode="user"} 0.68
node_cpu{cpu="cpu0",mode="nice"} 0
node_cpu{cpu="cpu0",mode="system"} 3.04
node_cpu{cpu="cpu0",mode="idle"} 111.02
node_cpu{cpu="cpu0",mode="iowait"} 0.06
node_cpu{cpu="cpu0",mode="irq"} 0
node_cpu{cpu="cpu0",mode="softirq"} 0
node_cpu{cpu="cpu0",mode="steal"} 0
node_cpu{cpu="cpu0",mode="guest"} 0
node_cpu{cpu="cpu0",mode="guest_nice"} 0
# TYPE node_filefd_allocated gauge
node_filefd_allocated 288
# TYPE node_filefd_maximum gauge
node_filefd_maximum 50421
# TYPE node_forks counter
node_forks 3000
# TYPE node_intr counter
node_intr 12880
# TYPE node_load1 gauge
node_load1 0.18
# TYPE node_load15 gauge
node_load15 0.06
# TYPE node_load5 gauge
node_load5 0.15
# TYPE node_memory_MemTotal gauge
node_memory_MemTotal 5.20069e+08
# TYPE node_memory_MemFree gauge
node_memory_MemFree 4.76525e+08
# TYPE node_memory_MemAvailable gauge
node_memory_MemAvailable 4.60177e+08
# TYPE node_memory_Buffers gauge
node_memory_Buffers 5.1241e+06
# TYPE node_memory_Cached gauge
node_memory_Cached 1.21569e+07
# TYPE node_memory_SwapCached gauge
node_memory_SwapCached 0
# TYPE node_memory_Active gauge
node_memory_Active 1.6724e+07
# TYPE node_memory_Inactive gauge
node_memory_Inactive 5.86138e+06
# TYPE node_memory_Active_anon gauge
node_memory_Active_anon 5.36576e+06
# TYPE node_memory_Inactive_anon gauge
node_memory_Inactive_anon 40960
# TYPE node_memory_Active_file gauge
node_memory_Active_file 1.13582e+07
# TYPE node_memory_Inactive_file gauge
node_memory_Inactive_file 5.82042e+06
# TYPE node_memory_Unevictable gauge
node_memory_Unevictable 0
# TYPE node_memory_Mlocked gauge
node_memory_Mlocked 0
# TYPE node_memory_SwapTotal gauge
node_memory_SwapTotal 0
# TYPE node_memory_SwapFree gauge
node_memory_SwapFree 0
# TYPE node_memory_Dirty gauge
node_memory_Dirty 0
# TYPE node_memory_Writeback gauge
node_memory_Writeback 0
# TYPE node_memory_AnonPages gauge
node_memory_AnonPages 5.30842e+06
# TYPE node_memory_Mapped gauge
node_memory_Mapped 4.55885e+06
# TYPE node_memory_Shmem gauge
node_memory_Shmem 106496
# TYPE node_memory_Slab gauge
node_memory_Slab 1.14852e+07
# TYPE node_memory_SReclaimable gauge
node_memory_SReclaimable 4.33766e+06
# TYPE node_memory_SUnreclaim gauge
node_memory_SUnreclaim 7.14752e+06
# TYPE node_memory_KernelStack gauge
node_memory_KernelStack 1.2288e+06
# TYPE node_memory_PageTables gauge
node_memory_PageTables 802816
# TYPE node_memory_NFS_Unstable gauge
node_memory_NFS_Unstable 0
# TYPE node_memory_Bounce gauge
node_memory_Bounce 0
# TYPE node_memory_WritebackTmp gauge
node_memory_WritebackTmp 0
# TYPE node_memory_CommitLimit gauge
node_memory_CommitLimit 2.60035e+08
# TYPE node_memory_Committed_AS gauge
node_memory_Committed_AS 1.09896e+07
# TYPE node_memory_VmallocTotal gauge
node_memory_VmallocTotal 3.51844e+13
# TYPE node_memory_VmallocUsed gauge
node_memory_VmallocUsed 0
# TYPE node_memory_VmallocChunk gauge
node_memory_VmallocChunk 0
# TYPE node_memory_DirectMap4k gauge
node_memory_DirectMap4k 1.25174e+07
# TYPE node_memory_DirectMap2M gauge
node_memory_DirectMap2M 5.24288e+08
# TYPE node_network_receive_bytes gauge
node_network_receive_bytes{device="dummy0"} 0
node_network_receive_bytes{device="eth1"} 0
node_network_receive_bytes{device="lo"} 47486
node_network_receive_bytes{device="ifb0"} 0
node_network_receive_bytes{device="maratest"} 0
node_network_receive_bytes{device="teql0"} 0
node_network_receive_bytes{device="iface2"} 0
node_network_receive_bytes{device="ifb1"} 0
node_network_receive_bytes{device="eth0"} 1180
node_network_receive_bytes{device="br-lan"} 0
node_network_receive_bytes{device="eth2"} 12351
# TYPE node_network_receive_packets gauge
node_network_receive_packets{device="dummy0"} 0
node_network_receive_packets{device="eth1"} 0
node_network_receive_packets{device="lo"} 395
node_network_receive_packets{device="ifb0"} 0
node_network_receive_packets{device="maratest"} 0
node_network_receive_packets{device="teql0"} 0
node_network_receive_packets{device="iface2"} 0
node_network_receive_packets{device="ifb1"} 0
node_network_receive_packets{device="eth0"} 2
node_network_receive_packets{device="br-lan"} 0
node_network_receive_packets{device="eth2"} 94
# TYPE node_network_receive_errs gauge
node_network_receive_errs{device="dummy0"} 0
node_network_receive_errs{device="eth1"} 0
node_network_receive_errs{device="lo"} 0
node_network_receive_errs{device="ifb0"} 0
node_network_receive_errs{device="maratest"} 0
node_network_receive_errs{device="teql0"} 0
node_network_receive_errs{device="iface2"} 0
node_network_receive_errs{device="ifb1"} 0
node_network_receive_errs{device="eth0"} 0
node_network_receive_errs{device="br-lan"} 0
node_network_receive_errs{device="eth2"} 0
# TYPE node_network_receive_drop gauge
node_network_receive_drop{device="dummy0"} 0
node_network_receive_drop{device="eth1"} 0
node_network_receive_drop{device="lo"} 0
node_network_receive_drop{device="ifb0"} 0
node_network_receive_drop{device="maratest"} 0
node_network_receive_drop{device="teql0"} 0
node_network_receive_drop{device="iface2"} 0
node_network_receive_drop{device="ifb1"} 0
node_network_receive_drop{device="eth0"} 0
node_network_receive_drop{device="br-lan"} 0
node_network_receive_drop{device="eth2"} 0
# TYPE node_network_receive_fifo gauge
node_network_receive_fifo{device="dummy0"} 0
node_network_receive_fifo{device="eth1"} 0
node_network_receive_fifo{device="lo"} 0
node_network_receive_fifo{device="ifb0"} 0
node_network_receive_fifo{device="maratest"} 0
node_network_receive_fifo{device="teql0"} 0
node_network_receive_fifo{device="iface2"} 0
node_network_receive_fifo{device="ifb1"} 0
node_network_receive_fifo{device="eth0"} 0
node_network_receive_fifo{device="br-lan"} 0
node_network_receive_fifo{device="eth2"} 0
# TYPE node_network_receive_frame gauge
node_network_receive_frame{device="dummy0"} 0
node_network_receive_frame{device="eth1"} 0
node_network_receive_frame{device="lo"} 0
node_network_receive_frame{device="ifb0"} 0
node_network_receive_frame{device="maratest"} 0
node_network_receive_frame{device="teql0"} 0
node_network_receive_frame{device="iface2"} 0
node_network_receive_frame{device="ifb1"} 0
node_network_receive_frame{device="eth0"} 0
node_network_receive_frame{device="br-lan"} 0
node_network_receive_frame{device="eth2"} 0
# TYPE node_network_receive_compressed gauge
node_network_receive_compressed{device="dummy0"} 0
node_network_receive_compressed{device="eth1"} 0
node_network_receive_compressed{device="lo"} 0
node_network_receive_compressed{device="ifb0"} 0
node_network_receive_compressed{device="maratest"} 0
node_network_receive_compressed{device="teql0"} 0
node_network_receive_compressed{device="iface2"} 0
node_network_receive_compressed{device="ifb1"} 0
node_network_receive_compressed{device="eth0"} 0
node_network_receive_compressed{device="br-lan"} 0
node_network_receive_compressed{device="eth2"} 0
# TYPE node_network_receive_multicast gauge
node_network_receive_multicast{device="dummy0"} 0
node_network_receive_multicast{device="eth1"} 0
node_network_receive_multicast{device="lo"} 0
node_network_receive_multicast{device="ifb0"} 0
node_network_receive_multicast{device="maratest"} 0
node_network_receive_multicast{device="teql0"} 0
node_network_receive_multicast{device="iface2"} 0
node_network_receive_multicast{device="ifb1"} 0
node_network_receive_multicast{device="eth0"} 0
node_network_receive_multicast{device="br-lan"} 0
node_network_receive_multicast{device="eth2"} 0
# TYPE node_network_transmit_bytes gauge
node_network_transmit_bytes{device="dummy0"} 0
node_network_transmit_bytes{device="eth1"} 1340
node_network_transmit_bytes{device="lo"} 47486
node_network_transmit_bytes{device="ifb0"} 0
node_network_transmit_bytes{device="maratest"} 0
node_network_transmit_bytes{device="teql0"} 0
node_network_transmit_bytes{device="iface2"} 0
node_network_transmit_bytes{device="ifb1"} 0
node_network_transmit_bytes{device="eth0"} 2611
node_network_transmit_bytes{device="br-lan"} 1340
node_network_transmit_bytes{device="eth2"} 9392
# TYPE node_network_transmit_packets gauge
node_network_transmit_packets{device="dummy0"} 0
node_network_transmit_packets{device="eth1"} 10
node_network_transmit_packets{device="lo"} 395
node_network_transmit_packets{device="ifb0"} 0
node_network_transmit_packets{device="maratest"} 0
node_network_transmit_packets{device="teql0"} 0
node_network_transmit_packets{device="iface2"} 0
node_network_transmit_packets{device="ifb1"} 0
node_network_transmit_packets{device="eth0"} 18
node_network_transmit_packets{device="br-lan"} 10
node_network_transmit_packets{device="eth2"} 62
# TYPE node_network_transmit_errs gauge
node_network_transmit_errs{device="dummy0"} 0
node_network_transmit_errs{device="eth1"} 0
node_network_transmit_errs{device="lo"} 0
node_network_transmit_errs{device="ifb0"} 0
node_network_transmit_errs{device="maratest"} 0
node_network_transmit_errs{device="teql0"} 0
node_network_transmit_errs{device="iface2"} 0
node_network_transmit_errs{device="ifb1"} 0
node_network_transmit_errs{device="eth0"} 0
node_network_transmit_errs{device="br-lan"} 0
node_network_transmit_errs{device="eth2"} 0
# TYPE node_network_transmit_drop gauge
node_network_transmit_drop{device="dummy0"} 0
node_network_transmit_drop{device="eth1"} 0
node_network_transmit_drop{device="lo"} 0
node_network_transmit_drop{device="ifb0"} 0
node_network_transmit_drop{device="maratest"} 0
node_network_transmit_drop{device="teql0"} 0
node_network_transmit_drop{device="iface2"} 0
node_network_transmit_drop{device="ifb1"} 0
node_network_transmit_drop{device="eth0"} 0
node_network_transmit_drop{device="br-lan"} 0
node_network_transmit_drop{device="eth2"} 0
# TYPE node_network_transmit_fifo gauge
node_network_transmit_fifo{device="dummy0"} 0
node_network_transmit_fifo{device="eth1"} 0
node_network_transmit_fifo{device="lo"} 0
node_network_transmit_fifo{device="ifb0"} 0
node_network_transmit_fifo{device="maratest"} 0
node_network_transmit_fifo{device="teql0"} 0
node_network_transmit_fifo{device="iface2"} 0
node_network_transmit_fifo{device="ifb1"} 0
node_network_transmit_fifo{device="eth0"} 0
node_network_transmit_fifo{device="br-lan"} 0
node_network_transmit_fifo{device="eth2"} 0
# TYPE node_network_transmit_colls gauge
node_network_transmit_colls{device="dummy0"} 0
node_network_transmit_colls{device="eth1"} 0
node_network_transmit_colls{device="lo"} 0
node_network_transmit_colls{device="ifb0"} 0
node_network_transmit_colls{device="maratest"} 0
node_network_transmit_colls{device="teql0"} 0
node_network_transmit_colls{device="iface2"} 0
node_network_transmit_colls{device="ifb1"} 0
node_network_transmit_colls{device="eth0"} 0
node_network_transmit_colls{device="br-lan"} 0
node_network_transmit_colls{device="eth2"} 0
# TYPE node_network_transmit_carrier gauge
node_network_transmit_carrier{device="dummy0"} 0
node_network_transmit_carrier{device="eth1"} 0
node_network_transmit_carrier{device="lo"} 0
node_network_transmit_carrier{device="ifb0"} 0
node_network_transmit_carrier{device="maratest"} 0
node_network_transmit_carrier{device="teql0"} 0
node_network_transmit_carrier{device="iface2"} 0
node_network_transmit_carrier{device="ifb1"} 0
node_network_transmit_carrier{device="eth0"} 0
node_network_transmit_carrier{device="br-lan"} 0
node_network_transmit_carrier{device="eth2"} 0
# TYPE node_network_transmit_compressed gauge
node_network_transmit_compressed{device="dummy0"} 0
node_network_transmit_compressed{device="eth1"} 0
node_network_transmit_compressed{device="lo"} 0
node_network_transmit_compressed{device="ifb0"} 0
node_network_transmit_compressed{device="maratest"} 0
node_network_transmit_compressed{device="teql0"} 0
node_network_transmit_compressed{device="iface2"} 0
node_network_transmit_compressed{device="ifb1"} 0
node_network_transmit_compressed{device="eth0"} 0
node_network_transmit_compressed{device="br-lan"} 0
node_network_transmit_compressed{device="eth2"} 0
# TYPE node_procs_blocked gauge
node_procs_blocked 0
# TYPE node_procs_running gauge
node_procs_running 1
# TYPE node_time counter
node_time 1.48854e+09
# TYPE node_uname_info gauge
node_uname_info{domainname="(none)",machine="x86_64",nodename="lede-dev",release="4.9.10",sysname="Linux",version="#0 SMP Fri Feb 17 08:44:23 2017"} 1
````

**Comment by danrl** [comment]
How far are we from getting this merged? Is there anything else I can do to help making the merging decision?

**Comment by danrl** [comment]
This work continues here: https://github.com/openwrt/packages/pull/4331
